pkg/proxy/iptables/proxier.go

```diff
@@ -328,6 +333,7 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
 	// Remove services missing from the update.
 	for name := range proxier.serviceMap {
 		if !activeServices[name] {
+			proxier.recorder.Eventf(proxier.nodeRef, "WatchForServiceFailed", "Watch failed, now stopping service %q", name)
```

Comment:
If a failed watch is causing Services to be removed that is a SERIOUS bug.
Watch is supposed to fail every now and then.

On Wed, Sep 23, 2015 at 6:00 AM, aveshagarwal <notifications@github.com>
wrote:

> In pkg/proxy/iptables/proxier.go
> <https://github.com/kubernetes/kubernetes/pull/14247#discussion_r40199551>
> :
>
> > @@ -328,6 +333,7 @@ func (proxier *Proxier) OnServiceUpdate(allServices []api.Service) {
> >  	// Remove services missing from the update.
> >  	for name := range proxier.serviceMap {
> >  		if !activeServices[name] {
> > +			proxier.recorder.Eventf(proxier.nodeRef, "WatchForServiceFailed", "Watch failed, now stopping service %q", name)
>
> @thockin <https://github.com/thockin> Thanks for your feedback. I agree
> with your comments and when I look at it (WatchForServiceFailed) and others
> again, it seems to me that it might be misleading to say Failed, although I
> am explaining below why I chose those strings.
>
> This condition occurs as you said that when a service or endpoints is
> removed. That means if as soon as the service or endpoints is not part of
> the update set the "first time", they are removed. However, my
> understanding was that there might be 2 reasons that this (service or
> endpoints not part of update set) occurred : 1) Removed explicitly, 2)
> Watch failed. Since, as per my (may be flawed) understanding, it is not
> possible to distinguish between 2 reasons at this stage, I chose to go with
> these message strings.
>
> So my question is that is it possible at this stage to distinguish between
> explicit removal and watch failed that caused service or endpoints not part
> of the update set?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/14247/files#r40199551>.
>